File: lib/features/prontuario/domain/entities/odontogram.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'odontogram.freezed.dart';
part 'odontogram.g.dart';

@freezed
class ToothCondition with _$ToothCondition {
  const factory ToothCondition({
    required int toothNumber,
    required List<ToothSurface> surfaces,
    required ConditionType condition,
    String? observation,
    DateTime? lastUpdate,
  }) = _ToothCondition;

  factory ToothCondition.fromJson(Map<String, dynamic> json) => _$ToothConditionFromJson(json);
}

enum ToothSurface {
  @JsonValue('Mesial') mesial,
  @JsonValue('Distal') distal,
  @JsonValue('Occlusal') occlusal,
  @JsonValue('Buccal') buccal,
  @JsonValue('Lingual') lingual,
  @JsonValue('Palatal') palatal,
  @JsonValue('Root') root
}

enum ConditionType {
  @JsonValue('Healthy') healthy,
  @JsonValue('Decayed') decayed,
  @JsonValue('Restored') restored,
  @JsonValue('Missing') missing,
  @JsonValue('Implant') implant,
  @JsonValue('Endodontic') endodontic,
  @JsonValue('Prosthesis') prosthesis
}

@freezed
class Odontogram with _$Odontogram {
  const factory Odontogram({
    required String id,
    required String patientId,
    required List<ToothCondition> teeth,
    required DateTime updatedAt,
    required String updatedBy,
  }) = _Odontogram;

  factory Odontogram.fromJson(Map<String, dynamic> json) => _$OdontogramFromJson(json);

  Map<String, dynamic> toJson() => _$$OdontogramImplToJson(this as _$OdontogramImpl);
}

File: backend/DentalClinic.Core/Domain/Entities/Odontogram.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace DentalClinic.Core.Domain.Entities;

/// <summary>
/// Representa o estado clínico da arcada dentária de um paciente.
/// </summary>
public sealed class Odontogram : Entity
{
    public Guid PatientId { get; private set; } // Alterado de int para Guid
    public Patient Patient { get; private set; } = null!;

    public List<ToothCondition> Teeth { get; private set; } = new();

    public Guid UpdatedBy { get; private set; }

    private Odontogram() { }

    public static Odontogram Create(Guid patientId, Guid userId)
    {
        return new Odontogram
        {
            PatientId = patientId,
            UpdatedBy = userId,
            Teeth = new List<ToothCondition>()
        };
    }

    public void UpdateTooth(int toothNumber, ConditionType condition, List<ToothSurface> surfaces, string? observation, Guid userId)
    {
        var existing = Teeth.FirstOrDefault(t => t.ToothNumber == toothNumber);
        if (existing != null)
        {
            Teeth.Remove(existing);
        }

        Teeth.Add(new ToothCondition(toothNumber, surfaces, condition, observation));
        UpdatedBy = userId;
        UpdatedAt = DateTime.UtcNow;
    }
}

public record ToothCondition(int ToothNumber, List<ToothSurface> Surfaces, ConditionType Condition, string? Observation);

public enum ToothSurface
{
    Mesial, Distal, Occlusal, Buccal, Lingual, Palatal, Root
}

public enum ConditionType
{
    Healthy, Decayed, Restored, Missing, Implant, Endodontic, Prosthesis
}
